proposals/0000-unsaturated-type-families.rst

```diff
@@ -309,17 +426,147 @@ the user needs to provide either a concrete matchability or a wildcard before
 preferable to write ``qux @Id`` instead.
+
 The `explicit specificity <https://github.com/ghc-proposals/ghc-proposals/pull/99>`_
-proposal would make this possible.
+feature greatly improves the usability of unsaturated type families, as now the signature
+can be written as ::
+
+   qux :: forall {m} (f :: * -> @m *) a. f a -> f a
+
+Standalone Kind Signatures
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``StandaloneKindSignatures`` feature has the largest user-facing interaction
+surface with the current proposal. Before ``StandaloneKindSignatures``, the kind
+of a type family was never written in the source language, and was only hinted at by
+GHCi. Today this is no longer the case ::
+
+  type Id :: Type -> Type
+  type family Id a where
+    Id a = a
+
+The meaning of this kind signature is ``type Id :: Type -> @U Type``. However,
+requiring the user to spell out the full signature incurs a considerable amount
+of mental burden, so instead we propose to infer it from the definition.
+
+Another example ::
+
+  type Foo :: Type -> Type
+  type family Foo where
+    Foo = Maybe
+
+  type Bar :: Type -> Type
+  type family Bar where
+    Bar = Id
+
+will get inferred to be ``Foo :: Type -> @M Type`` and ``Bar :: Type -> @U Type`` respectively.
+We can also infer polymorphism, for example ::
+
+  type Map :: (a -> b) -> [a] -> [b]
+  type family Map f xs where
+    Map f       '[] =             '[]
+    Map f (x ': xs) = f x ': Map f xs
+
+can be inferred the kind ``forall {t :: Matchability} {k :: Type} (a b :: k). (a -> @t b) -> @U [a] -> @U [b]``.
+
+When such inference is not possible, we propose to use a defaulting strategy
+that is in line with existing behaviour. For example, in the case of open type
+families ::
+
+  type Open :: Type -> Type
+  type family Open
+
+we can not tell up front which arrow the user meant, so we default it to mean
+matchable. Doing so disallows defining ``type instance Open = Id``, but such
+instances are already disallowed today due to the saturation restriction.
+
+The proposed solution requires some inference to happen even in checking
+mode. Kind inference is notoriously hard due to issues like kind-indexed type
+families and polymorphic recursion. However, we propose only to infer
+matchabilities and not arbitary kinds, which is a conservative extension of the
+status quo. In particular, we never infer non-parametric polymorphism so the following
+is rejected ::
+
+  type FooBar :: Type -> Type
+  type family FooBar where
+    FooBar = Maybe
+    FooBar = Id
+
+To be accepted, an annotation is required::
+
+  type FooBar' :: Type -> @m Type
+  type family FooBar' where
+    FooBar' = Maybe
+    FooBar' = Id
+
+It means that in type families, users only need to provide an explicit signature
+if they desire the matchability-indexed definition, which we predict is going to
+be restricted to a small fraction of advanced users.
+
+CUSKs
+~~~~~
+
+In the proposed solution the requirements to have a CUSK_ changes.
```

Comment:
Is there a need for this? We could just default `m` to `U` for type families / type synonyms, and `M` for data families / data types / classes.

Reply:
We could do defaulting here, but that would mean we can infer less polymorphism.
```haskell
type family Foo (f :: Type -> Type) :: Type -> Type where
  Foo f = f
```
can be polymorphic. Defaulting to unmatchable here is a breaking change, defaulting to matchable is potentially confusing, inferring
```
Foo :: (Type -> @m Type) -> Type -> @m Type
```
is the sweet spot to me. Does that answer your question?

Reply:
> Defaulting to unmatchable here is a breaking change

But making the condition for having a CUSK more demanding is also a breaking change. CUSKs are a way of making mutually recursive definitions work (just as SAKSs are), and if, under the new scheme, fewer definitions are considered to have a CUSK, then there will be fewer programs that type check.

Reply:
This is true. I have considered an alternative to it (in the alternatives section) which would not be too hard to implement. Perhaps I should move that into the main body of the proposal and expand on the strategy?

Reply:
Once again thanks for raising this point. I've updated the proposal with a backwards-compatible CUSK strategy (which I've also implemented).